Firestore/core/src/firebase/firestore/model/field_value.cc

```diff
@@ -36,132 +36,395 @@ namespace firebase {
 namespace firestore {
 namespace model {
 
+using BaseValue = FieldValue::BaseValue;
 using Type = FieldValue::Type;
 
 using util::Compare;
 using util::ComparisonResult;
 
-std::string ServerTimestamp::ToString() const {
-  std::string time = local_write_time.ToString();
-  return absl::StrCat("ServerTimestamp(local_write_time=", time, ")");
+template <typename T>
+static const T& Cast(const std::shared_ptr<BaseValue>& rep) {
```

Comment:
Optional: `Cast` is a nice abstraction, but almost all code below only uses `Cast` to access the value. Would it be useful to define something like:
```cpp
template <typename T>
auto As(const std::shared_ptr<BaseValue>& rep) -> decltype(Cast<T>(rep).value()) {
  return Cast<T>(rep).value();
}
// ...
return As<IntegerValue>(*this);
```
?

Reply:
In my view, the win is having one versus two function calls, not necessarily the character count. However, I'm totally fine if you leave as is.